File: python/day09.py
instructions = [int(n) for n in open("inputs/09.txt").readlines()]

part_1 = lambda instr: next(num for i, num in enumerate(instr[25:])
                            if not any (num-m in instr[i:25+i]
                                        for m in instr[i:25+i]))

part_2 = (lambda instr, goal, lo=0, hi=2:
          min(r) + max(r) if sum(r := instr[lo:hi]) == goal
          else part_2(instr, goal, lo, hi+1) if sum(r) < goal
          else part_2(instr, goal, lo+1, hi))

p1 = part_1(instructions)
print(p1)
print(part_2(instructions, p1))

File: nim/day09.nim
import aoc

const PreambleLength = 25

func isSum(input: seq[int], i: int): bool =
  for a in i-PreambleLength ..< i:
    for b in a+1 ..< i:
      if input[a] + input[b] == input[i]:
        return true

func part1(input: seq[int]): int =
  for i in PreambleLength ..< input.len:
    if not isSum(input, i):
      return input[i]

func part2(input: seq[int], goal: int): int =
  var lo = 0
  var hi = 1
  var total = input[lo] + input[hi]
  while total != goal:
    if total < goal:
      inc hi
      total += input[hi]
    else:
      total -= input[lo]
      inc lo
  return min(input[lo..hi]) + max(input[lo..hi])

let input = parseIntSeq "inputs/09.txt"
let p1 = part1(input)

echo p1
echo part2(input, p1)
